eth/filters/filter.go

```diff
@@ -95,43 +95,38 @@ func newFilter(sys *FilterSystem, addresses []common.Address, topics [][]common.
 // Logs searches the blockchain for matching log entries, returning all from the
 // first block that contains matches, updating the start of the filter accordingly.
 func (f *Filter) Logs(ctx context.Context) ([]*types.Log, error) {
-	// If we're doing singleton block filtering, execute and return
-	if f.block != nil {
-		header, err := f.sys.backend.HeaderByHash(ctx, *f.block)
-		if err != nil {
-			return nil, err
-		}
-		if header == nil {
-			return nil, errors.New("unknown block")
-		}
-		return f.blockLogs(ctx, header)
+	logChan, errChan, err := f.logsAsync(ctx)
+	if err != nil {
+		return nil, err
 	}
-	// Short-cut if all we care about is pending logs
-	if f.begin == rpc.PendingBlockNumber.Int64() {
-		if f.end != rpc.PendingBlockNumber.Int64() {
-			return nil, errors.New("invalid block range")
+
+	var logs []*types.Log
+LOOP:
+	for {
+		select {
+		case log := <-logChan:
+			logs = append(logs, log)
+		case ierr := <-errChan:
+			err = ierr
+			break LOOP
 		}
-		return f.pendingLogs()
-	}
-	// Figure out the limits of the filter range
-	header, _ := f.sys.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
-	if header == nil {
-		return nil, nil
 	}
-	var (
-		err     error
-		head    = header.Number.Int64()
-		pending = f.end == rpc.PendingBlockNumber.Int64()
-	)
+	return logs, err
+}
+
+// logsAsync retrieves logs that match the filter criteria asynchronously,
+// it creates and returns two channels: one for delivering log data, and one for reporting errors.
+func (f *Filter) logsAsync(ctx context.Context) (chan *types.Log, chan error, error) {
```

Comment:
We can simplify this function more: please rename this `logsRangeAsync` which takes in addition to ctx the begin and end block numbers (resolved already from special names) and only handles log retrieval for range.

Then we can move single block filtering (i.e. `f.block != nil`) and pending block handling to `Logs()`. In subscription we mostly deal with ranges anyway.

Reply:
thanks, please take another look